Publish the RAND_DRBG API

mspncp opened the pull request:
As announced long time ago in my personal roadmap #4403, I would like to publish the RAND_DRBG (and RAND_POOL) functions as officially supported API for the OpenSSL 1.1.1 release. 

Documentation for the new API is work in progress on  #5461.

mspncp commented:
Failure of Travis job [16596.19](https://travis-ci.org/openssl/openssl/jobs/346060740) is unrelated: `apt-get install failed`.

kroeckx commented:
I've restarted that job

kroeckx commented:
Do all of those functions need to move to the public headers? Don't the .num files need to be updated?

mspncp commented:
I will recheck this. Some of the function could probably be omitted from the public headers. Most notably definitions like 
```
#define RAND_DRBG_get_ex_new_index(l, p, newf, dupf, freef) \
    CRYPTO_get_ex_new_index(CRYPTO_EX_INDEX_DRBG, l, p, newf, dupf, freef)
int RAND_DRBG_set_ex_data(RAND_DRBG *drbg, int idx, void *arg);
void *RAND_DRBG_get_ex_data(const RAND_DRBG *drbg, int idx);
```

However, most of the RAND_DRBG and RAND_POOL functions are needed. The symbols are already in the export file.

```
msp@msp-vmware:~/src/openssl/util$ grep RAND_DRBG *.num
libcrypto.num:RAND_DRBG_instantiate                   4292	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_uninstantiate                 4293	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_set                           4295	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_set_callbacks                 4296	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_new                           4297	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_set_reseed_interval           4298	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_free                          4299	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_generate                      4300	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_reseed                        4301	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_set_ex_data                   4302	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_get_ex_data                   4303	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_get0_public                   4319	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_get0_private                  4329	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_get0_master                   4394	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_set_reseed_time_interval      4395	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_bytes                         4445	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_DRBG_secure_new                    4446	1_1_1	EXIST::FUNCTION:
```

```
msp@msp-vmware:~/src/openssl/util$ grep RAND_POOL *.num
libcrypto.num:RAND_POOL_new                           4351	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_free                          4352	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_buffer                        4353	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_detach                        4354	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_entropy                       4355	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_length                        4356	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_entropy_available             4357	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_entropy_needed                4358	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_bytes_needed                  4359	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_bytes_remaining               4360	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_add                           4361	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_add_begin                     4362	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_add_end                       4363	1_1_1	EXIST::FUNCTION:
libcrypto.num:RAND_POOL_acquire_entropy               4364	1_1_1	EXIST::FUNCTION:
```

kroeckx commented:
Some of the functions like RAND_DRBG_bytes() currently document that a lock should be hold when calling it. I think that we prefer that for the external API, the lock should not be taken external and that if needed the function should take the lock itself.

kroeckx commented:
Why do we need the RAND_POOL externally?

mspncp commented:
Everybody who needs to replace the default entropy callback using `RAND_DRBG_set_callbacks()` will find the `RAND_POOL` API very helpful to implement the `get_entropy()` callback. It's not a must-have, but it completes the API nicely.

*Note*: Replacing the default entropy callbacks is _not_ a hypothetical scenario, it is precisely what my company and I are currently doing in OpenSSL 1.0.2, using the OpenSSL FIPS DRBG. We have a little hardware device which has little entropy sources, so the requirement was to use the smartcard's random generator as entropy source, from which we reseed the DRBG on _every_ generate request, such that effectively the entire entropy comes from the smartcard, but is mixed by the AES-CTR DRBG. The /dev/urandom device is used only as additional input. We are stuck currently with OpenSSL 1.0.2, because we are using the FIPS DRBG, and the main force which has driven my work on the new DRBG API is to make the transition to OpenSSL 1.1+ possible for us.

richsalz commented:
For the exdata stuff, just add an entry to the exdata manpage.

kroeckx commented:
I think to judge that if this is an API we want to make public we need to understand the API, and I think for most of the other OMC members the easiest way for that is to have it properly documented.

I think there are at least some things I'm not happy with that we might want to change before the release.

mspncp commented:
I was thinking of your concerns, and maybe you are right. Maybe it is too early to publish the `RAND_POOL` api for collecting entropy. @paulidale startet together with us in #4394 to rethink the process of collecting the randomness from various entropy sources, but we never got very far with it. The current strategy is simply: use the first suitable method available and obtain as much entropy as you can get of it. So maybe we should leave the `RAND_POOL`api private until know more about how to deal with the questions in #4394.

mspncp commented:
Rebased to the tip of master. No changes.

mspncp commented:
The RAND_POOL API is now private agaub, its function prototypes were moved to "internal/rand_int.h".

mspncp commented:
Rebased without changes to pick up `RAND_DRBG_set_reseed_defaults()` from #5576.

levitte commented on util/libcrypto.num:
```diff
@@ -4402,20 +4402,6 @@ EVP_PKEY_set1_engine                    4347	1_1_0g	EXIST::FUNCTION:ENGINE
-RAND_POOL_add_begin                     4362	1_1_1	EXIST::FUNCTION:
-RAND_POOL_add_end                       4363	1_1_1	EXIST::FUNCTION:
-RAND_POOL_acquire_entropy               4364	1_1_1	EXIST::FUNCTION:
 OPENSSL_sk_new_reserve                  4365	1_1_1	EXIST::FUNCTION:
```
We should remember to renumber the symbols following.  To be noted, there are two symbols marked 1_1_0h, a version that hasn't been released yet, so we will need to synchronise the 1.1.0 `util/libcrypto.num` accordingly.

If noone else wants to deal with that, I can.

mspncp replied:
With pleasure, I just didn't dare to renumber existing numbers. I can do the renumbering, if you tell me which rules I need to obey. For example:
1. Is there a rule that forbids renumbering exports which already exist in stable branches?
2. Must the number of a symbol be identical in all branches?
3. What about the symbols marked 1_1_0h? Do they have to be moved towards the beginning, in front of all symbols that are new in 1.1.1?

kroeckx replied:
You can't renumber anything in the released version, which is as good as the same as in a stable branch. The numbers should match in those branches that have the same soname, which means that with our version scheme since 1.0, the first 2 digits should match. 1.0.0, 1.0.1 and 1.0.2 all have the same soname. 1.1.0 and 1.1.1 also have the same soname, but different than the 1.0.X ones.

I guess the order of the numbers itself doesn't really matter as long as they're unique, but it might be useful have everything from 1.1.0 first, then 1.1.1. But that might not work somewhere in the future after the 1.1.1 release where we need to add a new function in 1.1.0 for security reasons.

levitte replied:
1. Yes, if those numbers are present in existing releases.  For the 1.1.0 branch, it means no touching those marked `1_1_0g` and older, but those marked `1_1_0h` are fine since 1.1.0h hasn't been released yet.
2. Those that co-exist both in master and in the 1.1.0 branch MUST be the same.  Since 1.1.0 broke the ABI, we also broke away all the numbers to be different from the 1.0.x series.  (*)
3. No, they only need to have the exact same number in the two branches.  Also, the last `1_1_0g` number is 4347, so you MUST NOT touch any number up to that one.

Me, I'd simply renumber 4365 and on down to become 4351 and on, and then take note of what happens with the `1_1_0h` numbers and adjust the 1.1.0 branch accordingly.

-----
(*) The reason for these numbers was Windows and is still VMS.  The way shared libraries work (and we still do that on Windows) is that each numbers works as an index to a transfer vector that points out the actual location of each named function or variable.  For ABI compatibility reasons, they must therefore be identical in all branches where we promise such compatibility.  Of course, the compatibility only works for an version upgrade, not for a version downgrade.
The next time we can throw all these numbers in the air is when we release 1.2.

mspncp replied:
Kurt and Richard, thanks for your detailed explanations!

P.S: Yeah, I‘m familiar with module definition files, I‘m a child of both worlds, Unix/Linux and Windows (although as a programmer, I prefer the former).

levitte replied:
> P.S: Yeah, I‘m familiar with module definition files, I‘m a child of both worlds, Unix/Linux and Windows (although as a programmer, I prefer the former).

Cool, good to know :-)

levitte commented:
There's this line in `include/openssl/ossl_typ.h`, shouldn't it be removed as well?

``` C
include/openssl/ossl_typ.h:typedef struct rand_pool_st RAND_POOL;
```

mspncp commented:
Yes, you are right. I will fix it.

mspncp commented:
> There's this line in `include/openssl/ossl_typ.h`, shouldn't it be removed as well?

Done

levitte commented on crypto/rand/rand_lcl.h:
```diff
@@ -198,7 +200,7 @@ struct rand_drbg_st {
     RAND_DRBG_get_nonce_fn get_nonce;
     RAND_DRBG_cleanup_nonce_fn cleanup_nonce;
-};
+} RAND_DRBG;
```
Not sure why you're adding this typedef here, as it's already done in `ossl_typ.h`...  and obviously, that one is included first, or a compiler would probably have complained that the typedef doesn't exist when it hits the `RAND_DRBG *parent;` line...

levitte replied:
I'm sorry, I put that more harshly than I intended...

mspncp commented:
> I'm sorry, I put that more harshly than I intended...

Well, it didn't sound harsh to me at all. Your question is justified, it deserves an answer. When I made the change I almost anticipated that somebody would question this unnecessary and 'useless' change. So here comes my answer for you and beware, it is _really long_. I hope you'll have the patience to read it until the end.

The short version is that the asymmetry between the following three definitions disturbed me: While `rand_drbg_method_st/RAND_DRBG_METHOD` and `rand_drbg_ctr_st/RAND_DRBG_CTR` were defined as typedef-structs,
```
typedef struct rand_drbg_method_st {
    RAND_DRBG_instantiate_fn instantiate;
    ...
} RAND_DRBG_METHOD;

typedef struct rand_drbg_ctr_st {
    AES_KEY ks;
    ...
} RAND_DRBG_CTR;
```

`rand_drbg_st` was defined only as a struct, although it's actually used later on like a typedef-struct.
 
```
struct rand_drbg_st {
    CRYPTO_RWLOCK *lock;
};
```

The longer explanation is: I think it makes sense to emphasize not only in the declaration but also in the definition that this is a typedef-struct and that `rand_drbg_st` and `RAND_DRBG` belong together. Now, before anyone intervenes: I know that a "typedef-struct" is not a C language construct. It's only a programming pattern that fixes a deficiency of the original C language design (which was fixed in C++), namely that you can't omit the `struct` when declaring variables of a structured type.

> Not sure why you're adding this typedef here, as it's already done in ossl_typ.h... and obviously, that one is included first, or a compiler would probably have complained that the typedef doesn't exist when it hits the RAND_DRBG *parent; line...

To argue that it is unnecessary here because the typedef is declared elsewhere is correct, but not the whole story. Explicit is always better, so it makes sense to define a typedef-struct using the same syntax as when declaring it, to emphasize the fact that it is used like a typedef-struct (i.e, the structure tag is a mostly useles appendix, like the human one). The syntactic analogy between declaration and definition is a common principle in C an the compiler explicitely supports this type of redundancy.

```
/* forward declaration of a typedef-struct */
typedef foo_st FOO;

...

/* definition of a typedef-struct */
typedef foo_st
{
   ...
} FOO;
```

Nobody would come up with the idea to omit the parameter list from a function's definition, only because it is already present in the function's declaration. And for the same reason it makes sense to emphasize the intent of using a struct as a typedef-struct by repeating it in its definition.

levitte commented:
Hmmm, ok, your reasoning makes sense.  It's not our usual pattern, but on the other hand, it doesn't hurt, and could set a precedent for us all.

> @levitte if you still prefer that I revert the typedef, no problem. Just let me know.

Nah :-)

levitte commented:
Travis failures point out that you didn't notice that some of the `util/*.num` numbers are duplicated.  This is happening when certain entries are treated differently on different platforms.

A simple `make update` fixes the immediate problem.  It does leave a few small gaps in the transfer vector, but we can live with that.  That is, unless you absolutely want to be meticulous about it ;-)

mspncp commented:
You caught me: _Of course I am meticulous!_  ;-)

levitte commented:
> You caught me: _Of course I am meticulous!_  ;-)

I kinda suspected that ;-)

levitte commented:
@mspncp, you're gonna hate this...  from Travis:

```
ccache gcc  -I. -Icrypto/include -Iinclude -fPIC -pthread -m64 -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wmissing-prototypes -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DDSO_DLFCN -DHAVE_DLFCN_H -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-1.1\"" -DNDEBUG  -MMD -MF crypto/rand/drbg_ctr.d.tmp -MT crypto/rand/drbg_ctr.o -c -o crypto/rand/drbg_ctr.o crypto/rand/drbg_ctr.c
In file included from crypto/rand/drbg_ctr.c:15:0:
crypto/rand/rand_lcl.h:203:3: error: redefinition of typedef 'RAND_DRBG' [-Werror=pedantic]
 } RAND_DRBG;
   ^
In file included from include/openssl/crypto.h:25:0,
                 from crypto/rand/drbg_ctr.c:12:
include/openssl/ossl_typ.h:117:29: note: previous declaration of 'RAND_DRBG' was here
 typedef struct rand_drbg_st RAND_DRBG;
                             ^
cc1: all warnings being treated as errors
```

mspncp commented:
Sigh! I'll revert...

```
In file included from crypto/rand/drbg_ctr.c:15:0:
crypto/rand/rand_lcl.h:203:3: error: redefinition of typedef 'RAND_DRBG' [-Werror=pedantic]
 } RAND_DRBG;
   ^
In file included from include/openssl/crypto.h:25:0,
                 from crypto/rand/drbg_ctr.c:12:
include/openssl/ossl_typ.h:117:29: note: previous declaration of 'RAND_DRBG' was here
 typedef struct rand_drbg_st RAND_DRBG;
                             ^
cc1: all warnings being treated as errors
```

levitte commented:
Also, you should take a closer look at what happens with Appveyor.  Almost all the test recipes are going belly up...

mspncp commented:
We noticed it independently :-)

mspncp commented:
I will. But currently, I have to do some work for my other employer...

mspncp commented:
Marked it WIP for the moment.

mspncp commented:
Rebased without changes to kick the ci bots...

levitte commented on crypto/include/internal/rand_int.h:
```diff
@@ -15,8 +15,52 @@
+ */
+size_t RAND_POOL_acquire_entropy(RAND_POOL *pool);
+
+#endif
```
Now, one thing to think about here is, will the RAND_POOL API ever become public?  In that case, nothing more to say...  *However*, if the expectation is that it will always be internal, then I'd suggest that the function names be made lower case, like we normally do.

(note: some may argue that the functions should be made lower case until we decide to make them public.  My attitude is a bit more lax regard this, especially since that's not a written policy, but more of a "cultural" thing)

mspncp replied:
Currently it's not decided yet whether the RAND_POOL API will be published (perhaps in a modified form). So yes, no problem, I converted the API to lower case for the moment: b689902a88cf.

levitte commented on crypto/rand/rand_lcl.h:
```diff
@@ -107,6 +107,8 @@ typedef struct rand_drbg_ctr_st {
 } RAND_DRBG_CTR;
 
+/* forward declaration */
+typedef struct rand_pool_st RAND_POOL;
```
This seems unnecessary.  Just make sure to include `"internal/rand_int.h"` before including `"rand_lcl.h"`

My reasoning is that code copying, even one line, is bad and should be avoided if possible (it's not always possible, but I think that at least an effort should be made), because a few years down the line, fixes have been made in one spot but not the other, and no one notices for a long while...  This has happened before in numerous projects and will most certainly happen again, so let's avoid it when we can, yeah?

levitte replied:
Works with me :-)

levitte commented on crypto/rand/rand_unix.c:
```diff
@@ -12,6 +12,7 @@
 #include "internal/cryptlib.h"
 #include <openssl/rand.h>
 #include "rand_lcl.h"
+#include "internal/rand_int.h"
```
So re the double forward declaration of `RAND_POOL`, if you just move this one line up one step...

mspncp commented:
Rebased and squashed.  The unsquashed branch was pushed to  https://github.com/mspncp/openssl/tree/pr-publish-rand-drbg-api-unsquashed for reference.

@openssl/omc: Ready for final review.

levitte commented:
I think it's time you take it out of WIP

mspncp commented:
> I think it's time you take it out of WIP

Yup, I forgot about the title! Thanks for the reminder!

mspncp commented:
Rebased to resolve conflicts in libcrypto.num

mspncp commented:
Thanks for the green light! I'll leave it hanging around for a few more hours, to give others the opportunity to comment.

mspncp commented:
Merged!! Finally! This is the culminating point a project which started more or less accidentally by a mail from me to Rich on [openssl-dev in June 2017](https://mta.openssl.org/pipermail/openssl-dev/2017-June/009439.html). I would have never anticipated that this would become such a big involvement. Thanks to everybody who reviewed and participated!

Matthias

mattcaswell commented:
Awesome! Congrats @mspncp!!

mspncp commented:
Thanks! :smiley:

richsalz commented:
I feel like a proud papa :)
